### ISSUE TITLE
Deployment to Azure

### ISSUE THREAD

**Issue by sbwalker** [oqtane/oqtane.framework#545]
@ADefWebserver has been trying to document the process to deploy Oqtane to Azure but has run into some issues...

"I tried to create installation directions: https://github.com/oqtane/oqtane.framework/files/4670945/InstallingOqtane.docx. The install .zip package will not install on Azure. Even if I publish to Azure directly from Visual Studio it comes up as a blank page"

Let's try to get this figured out, as deployment to Azure will be critical for adoption of Oqtane.

To clarify, the goal is for someone to be able to take the "Install" package from the Releases page https://github.com/oqtane/oqtane.framework/releases/tag/v1.0.0 and install it on Azure. The "Install" package was created by performing a standard Publish in Visual Studio and zipping up the contents of the Publish folder.

**Comment by mikecasas** [comment]
I think I got it.  My oqtane site is running at:

https://oqtane1.azurewebsites.net/

**Comment by sbwalker** [comment]
what did you have to do which is different from the instructions in the doc created by Michael?

**Comment by mikecasas** [comment]
When I downloaded the zip files and look in appsettings.json, the file looks like this:

```
{
  "ConnectionStrings": {
    "DefaultConnection": "Data Source=(LocalDb)\\MSSQLLocalDB;AttachDbFilename=|DataDirectory|\\Oqtane-202005192135.mdf;Initial Catalog=Oqtane-202005192135;Integrated Security=SSPI;"
  },
  "Runtime": "Server",
  "Installation": {
    "DefaultAlias": "",
    "HostPassword": "",
    "HostEmail": "",
    "SiteTemplate": "",
    "DefaultTheme": "",
    "DefaultLayout": "",
    "DefaultContainer": ""
  }
}
```

I deleted the connection string settings (meaning set it to blank), and then re-pushed the appsettings file up to azure and it seamed to work.

I tried a lot of different things, so I think that is what worked. 

I guess maybe an old connection string was in there when the package was created???

**Comment by sbwalker** [comment]
You are correct, the appsettings.json file in the Install package had a DefaultConnection value in it. I have now removed it and replaced the package here:  https://github.com/oqtane/oqtane.framework/releases/tag/v1.0.0

**Comment by ADefWebserver** [comment]
> You are correct, the appsettings.json file in the Install package had a DefaultConnection value in it. I have now updated it to "" and replaced the package here: https://github.com/oqtane/oqtane.framework/releases/tag/v1.0.0

I just deployed that here:
https://blazor-oqtane.azurewebsites.net/

but the installer does not show :(

**Comment by sbwalker** [comment]
@mikecasas was able to install using your instructions so their must be some other issue he resolved in the process?

**Comment by ADefWebserver** [comment]
Nevermind. The problem was that I did not have web sockets turned on. It now works.
I will try to get the article updated in the next two days.

**Comment by sbwalker** [comment]
Good news! :)

**Comment by ADefWebserver** [comment]
Article posted:
**Installing Blazor Oqtane**
http://blazorhelpwebsite.com/Blog/tabid/61/EntryId/4377/Installing-Blazor-Oqtane.aspx

**Comment by sbwalker** [comment]
@mikecasas are you able to reproduce the issue #546 on your Azure deployment?

**Comment by mikecasas** [comment]
@sbwalker Yes, the admin Dashboard causes the Circuit error. But I was able to add the HtmlText module to a page successfully.

**Comment by sbwalker** [comment]
When you choose the Add or Edit Page options on the dashboard do they also produce a circuit error?

**Comment by sbwalker** [comment]
And is anything logged in the Event Log?

**Comment by ADefWebserver** [comment]
I looked in the event log in Azure and I saw nothing. I think we need support for detailed errors added to track things like this down: https://stackoverflow.com/questions/57514541/how-to-turn-on-circuitoptions-detailederrors

**Comment by mikecasas** [comment]
No Adding or editing a page causes the same error. Also trying to edit the HtmlText module causes the error also. The page will go into "Edit Mode" successfully, but when you click to edit the content, the same error message appears.

![error1](https://user-images.githubusercontent.com/11261164/83027255-ea059c80-9ffd-11ea-9dec-0956e24063d3.PNG)

**Comment by ADefWebserver** [comment]
> No Adding or editing a page causes the same error. 

Yes, I have the same behavior.

**Comment by jimspillane** [comment]
Here is the exception that I get.

```
2020-05-27 14:15:00.396 +00:00 [Error] Microsoft.AspNetCore.Components.Server.Circuits.CircuitHost: Unhandled exception in circuit 'tERWJGtij_dEMBbS85Ny6CEj0_WHT90sjD-L7dM-8_U'.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Oqtane.UI.SiteRouter.Refresh() in C:\Users\Shaun.Walker\source\repos\sbwalker\oqtane.framework\Oqtane.Client\UI\SiteRouter.razor:line 107
   at Oqtane.UI.SiteRouter.LocationChanged(Object sender, LocationChangedEventArgs args) in C:\Users\Shaun.Walker\source\repos\sbwalker\oqtane.framework\Oqtane.Client\UI\SiteRouter.razor:line 304
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_0(Object state)
   at Microsoft.AspNetCore.Components.Rendering.RendererSynchronizationContext.ExecuteSynchronously(TaskCompletionSource`1 completion, SendOrPostCallback d, Object state)
   at Microsoft.AspNetCore.Components.Rendering.RendererSynchronizationContext.<>c.<.cctor>b__23_0(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at Microsoft.AspNetCore.Components.Rendering.RendererSynchronizationContext.ExecuteBackground(WorkItem item)
2020-05-27 14:15:00.397 +00:00 [Debug] Microsoft.AspNetCore.Components.Server.Circuits.CircuitHost: About to notify client of an error in circuit 'tERWJGtij_dEMBbS85Ny6CEj0_WHT90sjD-L7dM-8_U'.
```

**Comment by R-Lamberson** [comment]
The circuit error is a blazor issue, can be caused by app service not being correctly configured but another issue related to routing - https://github.com/dotnet/aspnetcore/issues/13594
I added Application Insights on my environment to capture the client logging and I am seeing the following 
![image](https://user-images.githubusercontent.com/48865653/83035613-49b07780-9fff-11ea-8ae7-c4e333671c3f.png)

**Comment by sbwalker** [comment]
The error is coming from line 107 in SiteRouter - which obviously makes sense because there is no defensive checking for if alias is null. That being said, the bigger problem is related why the alias is null - and why it seems to only be happening in the Azure environment. 

105        alias = await AliasService.GetAliasAsync(path, lastsyncdate);
106        SiteState.Alias = alias; // set state for services
107        lastsyncdate = alias.SyncDate;

My hunch is that uri.LocalPath behaves differently in the Azure environment than it does locally - which is causing the problem. 

        var path = uri.LocalPath.Substring(1);

**Comment by jimspillane** [comment]
It might be an IIS thing filtering the URL.  The escaped `%2F` might be the issue.  

/1/api/Alias/name/admin%2Flog?sync=20200527145009641 
```
2020-05-27 14:38:26.541 +00:00 [Information] Microsoft.AspNetCore.Hosting.Diagnostics: Request finished in 1.4141ms 200 application/json; charset=utf-8
2020-05-27 14:38:26.597 +00:00 [Information] Microsoft.AspNetCore.Hosting.Diagnostics: Request starting HTTP/1.1 GET https://jimsoqtane.azurewebsites.net/~/api/Alias/name/admin/log?sync=20200527143827311  
2020-05-27 14:38:26.597 +00:00 [Trace] Microsoft.AspNetCore.HostFiltering.HostFilteringMiddleware: All hosts are allowed.
2020-05-27 14:38:26.597 +00:00 [Trace] Microsoft.AspNetCore.HttpsPolicy.HstsMiddleware: Adding HSTS header to response.
2020-05-27 14:38:26.597 +00:00 [Debug] Microsoft.AspNetCore.StaticFiles.StaticFileMiddleware: The request path /~/api/Alias/name/admin/log does not match a supported file type
2020-05-27 14:38:26.597 +00:00 [Debug] Microsoft.AspNetCore.Routing.Matching.DfaMatcher: 1 candidate(s) found for the request path '/~/api/Alias/name/admin/log'
2020-05-27 14:38:26.597 +00:00 [Debug] Microsoft.AspNetCore.Routing.Matching.DfaMatcher: Endpoint 'Fallback {*path:nonfile}' with route pattern '{*path:nonfile}' is valid for the request path '/~/api/Alias/name/admin/log'
2020-05-27 14:38:26.597 +00:00 [Debug] Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware: Request matched endpoint '/_Host'
```


Could a wildcard for {name} help?

```
// GET api/<controller>/name/xxx?sync=yyyyMMddHHmmssfff
        [HttpGet("name/{*name}")]
        public Alias Get(string name, string sync)
```

**Comment by R-Lamberson** [comment]
@sbwalker I can see the SiteRouter exception also, I am wondering if the _Blazor 503 is failure that is triggering the AliasService to return null to the SiteRouter

**Comment by sbwalker** [comment]
I think you are on to something Mike. A typical url for invoking the Admin Dashboard should be structured like:

http://www.domain.com/10/Index ( where the "10" is a module id the Admin Dashboard module )

The uri.LocalPath should return "/10/Index" and we are ignoring the first slash. So the value passed to the backend API would be:

"10/index"

However it is possible the slash is being encoded on Azure to %2F - which means the value would be:

"10%2FIndex"

And the API controller method is not Url decoding the value so it may not be able to find a match - and returning an Alias of null.

**Comment by sbwalker** [comment]
Even more strange... there is fallback logic in the AliasController in case an Alias is not found:

            if (alias == null && aliases.Any())
            {
                // use first alias if name does not exist
                alias = aliases.FirstOrDefault();
            }

This error would seem to indicate that the installation has no Alias records in the database. Is it possible for someone to access their database on Azure and look in the Alias table and share the contents?

**Comment by R-Lamberson** [comment]
@sbwalker here is the request being logged with the null reference exception 
![image](https://user-images.githubusercontent.com/48865653/83051126-2bed0d80-a013-11ea-87a7-23903e51f6ae.png)
The decode of the url works so not sure what the issue is

**Comment by R-Lamberson** [comment]
I can confirm there is a record in the alias table on my azure db

**Comment by mikecasas** [comment]
```
AliasId	Name	TenantId	SiteId	CreatedBy	CreatedOn	ModifiedBy	ModifiedOn
1	oqtane1.azurewebsites.net	1	1		2020-05-26 17:45:41.580		2020-05-26 17:45:54.607
2	oqtane1.azurewebsites.net	2	-1		2020-05-27 10:48:47.067		2020-05-27 10:48:47.067
```

**Comment by jimspillane** [comment]
I think the controller throws a 404 because the route does not match.  IIS might be unescaping the %2F and the controller is sent an unexpected route.  Kestrel might be more forgiving.

The [catchall parameters](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-3.1#route-template-reference) goes into it:

> The catch-all parameter escapes the appropriate characters when the route is used to generate a URL, including path separator / characters. For example, the route foo/{*path} with route values { path = "my/path" } generates foo/my%2Fpath. Note the escaped forward slash. To round-trip path separator characters, use the ** route parameter prefix. The route foo/{**path} with { path = "my/path" } generates foo/my/path.

**Comment by sbwalker** [comment]
@mikecasas you have duplicate alias Names in your Alias table - that is a bug but I don't know if its related to the issue we are discussing.

**Comment by mikecasas** [comment]
ok, yes I was surprised to see the two records.

**Comment by sbwalker** [comment]
@mikecasas it appears you have 2 tenant records as well - can you share the contents of that table?

**Comment by R-Lamberson** [comment]
so I am seeing a 503 on the response which is server unavailable - so the URL is good but there is something that is causing the Alias service to not respond

**Comment by mikecasas** [comment]
```
TenantId	Name	DBConnectionString	Version	CreatedBy	CreatedOn	ModifiedBy	ModifiedOn
1	Master	Data Source=<MyDb>.database.windows.net;Initial Catalog=-test;User ID=admin2;Password=MyPassword	1.0.0		2020-05-26 17:45:40.233		2020-05-26 17:45:54.643
2	Master	Data Source=<MyDb>.database.windows.net;Initial Catalog=-test;User ID=admin2;Password=MyPassword	1.0.0		2020-05-27 10:48:47.013		2020-05-27 10:48:47.013
```

**Comment by ADefWebserver** [comment]
> ok, yes I was surprised to see the two records.

I looked in my database on Azure and I only have one record:

`1 | Master | Data Source=server.database.windows.net;Initial Catalog=Oqtane_db;User   ID=username;Password=pass | 1.0.0 |   | 2020-05-26T21:22:15.7230000 | 2020-05-26T21:22:26.6200000
`

**Comment by jimspillane** [comment]
I got it to work changing _AliasController.cs_ from

```
        // GET api/<controller>/name/xxx?sync=yyyyMMddHHmmssfff
        [HttpGet("name/{name}")]
        public Alias Get(string name, string sync)
```

to

```
        // GET api/<controller>/name/xxx?sync=yyyyMMddHHmmssfff
        [HttpGet("name/{**name}")]
        public Alias Get(string name, string sync)
```

There might be other places that need the wildcard(**) treatment.

**Comment by R-Lamberson** [comment]
That works for me also

**Comment by sbwalker** [comment]
@jimspillane it looks like you located the issue... can you submit a PR? I am not sure if there are many other places which may have this issue as in general API parameters are integers rather than strings

**Comment by jimspillane** [comment]
I will submit the PR. 

I think you are right. The other places that it failed, like opening the editor, are fixed with this change too.

**Comment by sbwalker** [comment]
Thank you everyone for your help! I believe this is fixed in #548
